### PR TITLE
build: Exclude running device stream tests in prod pipelines

### PR DESCRIPTION
**Pull request by barustum**
## Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [ ] This pull-request is submitted against the `master` branch.

## Description of the changes

- Extract the device streaming unit tests into a separate test file
- Add a new test category for the device streaming tests
- Wrote logic in the build scripts to exclude the test category when running against the prod pipeline

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -274,7 +275,14 @@ try
             Write-Host -ForegroundColor Cyan "Unit Test execution"
             Write-Host
 
-            RunTests "Unit tests" -filterTestCategory "TestCategory=Unit" -framework $framework
+            $testCategory = "TestCategory=Unit";
+
+            if ($skipDeviceStreamTests)
```
Comment (drwill-ms) [inline build.ps1:280]: By definition, unit tests shouldn't hit the service. We don't seem to have other exclusions in this if block, so should this be removed?

Comment (drwill-ms) [reply build.ps1:280]: Or do the other exclusions belong here too?

File: iothub/device/tests/DeviceStreamingTests.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NSubstitute;
+
+namespace Microsoft.Azure.Devices.Client.Test
+{
+    [TestClass]
+    [TestCategory("DeviceStreaming")]
+    [TestCategory("Unit")]
```
Comment (drwill-ms) [inline iothub/device/tests/DeviceStreamingTests.cs:14]: These might not be unit tests. Perhaps they were just miscategorized E2E tests?

Comment (drwill-ms) [reply iothub/device/tests/DeviceStreamingTests.cs:14]: Otherwise, I don't see how we've put real service calls in West Central US, causing this problem.

Comment (drwill-ms) [reply iothub/device/tests/DeviceStreamingTests.cs:14]: According to the email, we should be looking for these test names:

> ModuleNotFoundTestSasl...
> DeviceClient_Amqp_TimesOutSasl_...

**Review by drwill-ms — CHANGES_REQUESTED** [review]
E2E tests need categorizing

File: build.ps1 (hunk 1 above)
Comment (barustum) [reply build.ps1:280]: Good catch. Removed this from the unittests section as they hit a fake endpoint (mock tests).

File: iothub/device/tests/DeviceStreamingTests.cs (hunk 2 above)
Comment (barustum) [reply iothub/device/tests/DeviceStreamingTests.cs:14]: me and @abhipsaMisra  are trying to figure out where these device names are coming from. We can't seem to find them in our code:
ModuleNotFoundTestSasl...
DeviceClient_Amqp_TimesOutSasl_...

Will report ASAP

Comment (barustum) [reply iothub/device/tests/DeviceStreamingTests.cs:14]: It seems like the tests that are problematic are the ones that start with E2E_DeviceStreamingTests_Sasl*, not as previously reported by the service team.

File: e2e/test/iothub/DeviceStreamingE2ETests.cs [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Azure.Devices.E2ETests.Helpers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 using ClientDeviceStreamingRequest = Microsoft.Azure.Devices.Client.DeviceStreamRequest;
 using ServiceDeviceStreamingRequest = Microsoft.Azure.Devices.DeviceStreamRequest;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
     [TestClass]
     [TestCategory("IoTHub-E2E")]
```
Comment (drwill-ms) [inline e2e/test/iothub/DeviceStreamingE2ETests.cs:20]: I think this should be broken out to E2E and IoTHub

**Review by drwill-ms — APPROVED** [review]

File: build.ps1 [hunk 4]
```diff
@@ -360,8 +366,14 @@ try
         # Override verbosity to display individual test execution.
         $oldVerbosity = $verbosity
         $verbosity = "normal"
+        $testCategory = "TestCategory=E2E"
+        
+        if ($skipDeviceStreamTests)
```
Comment (abhipsaMisra) [inline build.ps1:371]: I would add a note here stating that we are intentionally skipping e2e tests for batched/CI builds for certian scenarios.

File: vsts/determine_tests_to_run.ps1 [hunk 5]
```diff
@@ -39,6 +39,14 @@ function ShouldSkipPnPTests
 	return $false
 }
 
+function ShouldSkipDeviceStreamTests
+{
+	return ($env:SOURCE_BRANCH_NAME -and `
+		$env:SOURCE_BRANCH_NAME.toLower() -eq 'preview' -and `
+		$env:PIPELINE_ENVIRONMENT -and `
+		$env:PIPELINE_ENVIRONMENT.toLower() -eq 'prod')
```
Comment (abhipsaMisra) [inline vsts/determine_tests_to_run.ps1:47]: q - what are possible values for this env variable?

**Review by abhipsaMisra — APPROVED** [review]

**Review by timtay-microsoft — APPROVED** [review]

File: vsts/determine_tests_to_run.ps1 (hunk 5 above)
Comment (barustum) [reply vsts/determine_tests_to_run.ps1:47]: Good idea. Done.

**Pull request merged**